File: Sorting Algorithms/Quick Sort Algorithm/C/quick-sort-random-pivot.c
/* ===== ===== =====

Theory of Programming

Quick sort algorithm with random pivot selection
http://theoryofprogramming.com/2015/01/20/quick-sort-algorithm/
Code Contributor - Vamsi Sangam

===== ===== ===== */

#include <stdio.h>
#include <stdlib.h>

void quickSortRandomPivot(int arr[], int low, int high)
{
	if (low >= high) {
		// A single element is 
		// considered to be sorted
		return;
	}
	
	int mid = (rand() % (high - low)) + low;
	
	// Put the pivot at the last
	int temp = arr[mid];
	arr[mid] = arr[high];
	arr[high] = temp;
	
	int pivot = arr[high];
	int i = low;
	int j = high - 1;
	
	// The partition
	while (j >= i) {
		if (arr[j] < pivot && arr[i] > pivot) {
			temp = arr[i];
			arr[i] = arr[j];
			arr[j] = temp;
			
			++i;
			--j;
		} else if (arr[j] < pivot && arr[i] <= pivot) {
			++i;
		} else if (arr[j] >= pivot && arr[i] > pivot) {
			--j;
		} else if (arr[j] >= pivot && arr[i] <= pivot) {
			++i;
			--j;
		}
	}
	
	// Bring the pivot back to its
	// appropriate place
	temp = arr[i];
	arr[i] = arr[high];
	arr[high] = temp;
	
	quickSortRandomPivot(arr, low, i - 1);
	quickSortRandomPivot(arr, i + 1, high);
}

int main()
{
	int size, i;
	
	printf("Enter the size of the array -\n");
	scanf("%d", &size);
	
	int arr[size];
	
	printf("Enter %d Integers -\n", size);

	for (i = 0; i < size; ++i) {
		scanf("%d", &arr[i]);
	}
	
	quickSortRandomPivot(arr, 0, size - 1);
	
	printf("\nThe Sorted Array -\n");
	for (i = 0; i < size; ++i) {
		printf("%d ", arr[i]);
	}
	
	return 0;
}

File: Sorting Algorithms/Quick Sort Algorithm/Java/QuickSortRandomPivot.java
/* ===== ===== =====

Theory of Programming

Quick sort algorithm with random pivot selection
http://theoryofprogramming.com/2015/01/20/quick-sort-algorithm/
GitHub - https://github.com/VamsiSangam/theoryofprogramming
Code Contributor - Aditya Aggarwal (github.com/the-cybersapien, cybersapien.xyz)

===== ===== ===== */

import java.util.Scanner;

public class QuickSortRandomPivot {

    public static void main(String[] args) {
        // Initialize the Scanner Object to take input
        Scanner in = new Scanner(System.in);
        // Get the Size of the Array
        System.out.println("Enter the size of the array:");
        int size = in.nextInt();
        // Initialize the Array for sorting the elements
        int[] array = new int[size];
        // Get the data from user
        System.out.println("Enter the elements of the array: ");

        for (int i = 0; i < size; i++) {
            array[i] = in.nextInt();
        }

        // Call the quickSort function to sort the Array
        quickSortRandomPivot(array, 0, size - 1);

        // Print the Sorted Array
        System.out.println("Sorted Array: ");

        for (int a : array) {
            System.out.print(a + " ");
        }

        System.out.println();
    }

    private static void quickSortRandomPivot(int[] array, int low, int high){
        if (low >= high) {
            // A single element is
            // considered to be sorted
            return;
        }

        // Choose a random pivot using the random number generator
        // Additional brackets added here to ensure that there is no problems
        // while typecasting from double to int due to precedence
        int mid = (((int) (Math.random() * (high - low))) + low);

        // Put the pivot at the last
        int temp = array[mid];
        array[mid] = array[high];
        array[high] = temp;

        int pivot = array[high];
        int i = low;
        int j = high - 1;

        while (j >= i){
            if (array[j] < pivot && array[i] > pivot){
                temp = array[i];
                array[i] = array[j];
                array[j] = temp;

                i++;
                j--;
            } else if (array[j] < pivot && array[i] <= pivot){
                i++;
            } else if (array[j] >= pivot && array[i] > pivot){
                j--;
            } else if (array[j] >= pivot && array[i] <= pivot){
                i++;
                j--;
            }
        }

        // Bring the pivot back to its
        // appropriate place
        temp = array[i];
        array[i] = array[high];
        array[high] = temp;

        quickSortRandomPivot(array, low, i - 1);
        quickSortRandomPivot(array, i + 1, high);
    }
}